### PR TITLE
Make execution-statistics failures non-partial and observable (#117)

### LINKED ISSUE #117: price-level: Make execution-statistics failures non-partial and observable
## Overview

Make execution-statistics recording atomic on overflow and handle failures in `match_order`.

`record_execution` validates multiplication and timestamps first, but then mutates several independent atomics with fallible checked additions. If a later counter overflows, earlier counters have already advanced. `match_order` discards the returned error, so the trade succeeds with silently inconsistent statistics.

## Context

- Example: `orders_executed` and `quantity_executed` can advance before `value_executed` or `sum_waiting_time` rejects overflow.
- The queue/trade mutation has already committed when statistics are recorded, so returning a match error is not currently possible without a broader transaction policy.
- Statistics may be advisory under concurrent reads, but each accepted execution must contribute coherently to every aggregate.
- The fix must define deterministic overflow behavior and must not lose or duplicate trades.

## Tasks

- Introduce an all-or-nothing update strategy for the execution-statistics tuple or a documented saturating/wider-counter policy.
- Stop discarding `record_execution` failures in `match_order`.
- Define how a statistics overflow interacts with an already-committed maker/trade operation.
- Add boundary and concurrent tests that force overflow in each aggregate independently.
- Verify snapshot/statistics serialization exposes a coherent accepted-execution state.

## Acceptance criteria

- An accepted execution updates `orders_executed`, `quantity_executed`, `value_executed`, `last_execution_time`, and waiting-time aggregates coherently.
- An overflow cannot leave only a prefix of those fields mutated.
- `match_order` does not silently discard a statistics error; its chosen policy is explicit and documented.
- No fix path duplicates, removes, or reorders a committed trade.
- Normal matching retains FIFO/deterministic behavior and existing counter-to-queue invariants.
- Concurrent and boundary regression tests cover every fallible aggregate.
- `cargo fmt --all -- --check`, `cargo clippy --all-targets --all-features -- -D warnings`, `cargo test --all`, and `cargo build --release` pass.

## References

- `src/price_level/statistics.rs` (`record_execution`)
- `src/price_level/level.rs` (ignored result in `match_order`)
- Related statistics work: #66, #63

### PR DESCRIPTION
**Pull request by joaquinbejar**
## Summary

`record_execution` mutated several independent atomics with fallible checked
additions: a later overflow left the earlier counters advanced (partial
contribution), and `match_order` discarded the returned error — a trade
succeeded with silently inconsistent statistics. Statistics recording is now
all-or-nothing in the committed state, and a dropped execution is observable.

## Stack

- **Stack:** #118 → #114 → #116 → #111 → #113 → #120 → #119 → #115 → #117 → #112 (**this is 9/10**)
- **Base branch:** `stack/8-115-live-update`
- **Stacked on:** #128 — **the diff vs `main` is cumulative**; review against the base branch.

## Changes

- Additive aggregates commit via `fetch_update(checked_add)`; on any later
  failure the already-committed prefix rolls back with `fetch_sub` of exactly
  this call's contribution (the #111/#113 call-backed argument, documented as
  contingent on `reset()`'s new quiescence contract). `last_execution_time`
  stores only after every additive commit succeeds.
- `match_order` no longer discards the error: WARN log (the match is not
  aborted — the committed trade is never unwound) + new sticky
  `stats_degraded` flag on `PriceLevelStatistics`, queryable and carried
  through `Clone` / `Display` / `FromStr` / serde.
- **Snapshot compatibility preserved**: the flag serializes only when `true`,
  so a pre-#117 v2 package re-serializes byte-identically and its SHA-256
  still validates — old snapshots restore unchanged (guarded by an old-format
  compatibility test). No snapshot version bump needed.
- `record_order_added` / `record_order_removed` audited: single-atomic, no
  partiality possible.

## Technical decisions

- Validate-then-commit across 8 independent atomics is not linearizable
  without a lock; CAS-per-counter + prefix rollback keeps the path lock-free
  and matches the crate's rollback precedent. The transient prefix window
  (visible before rollback) is documented honestly — the guarantee is on the
  committed state.
- `architect` caught the HIGH checksum break (unconditional 9th field would
  have failed `validate()` on every pre-#117 package) — fixed with
  skip-when-false serialization. `concurrency-auditor` (0 P1) flagged the
  `reset()`-vs-rollback wrap (P2, unreachable today) — closed by contract:
  `reset()` documented quiescent-only. All LOW items applied (WARN level,
  Barrier in the concurrent test, config-specific exactly-K note, torn-read
  caveats on the average accessors).

## Public API impact

- New additive accessor `PriceLevelStatistics::stats_degraded()`; serde field
  emitted only when `true`. No signature changes, `lib.rs` untouched.

## Testing

- [x] Forced overflow per aggregate: no aggregate advances, flag set, `Err`
- [x] Barrier-controlled concurrent boundary: exactly K symmetric-headroom
  successes, `quantity_executed` / `value_executed` in lockstep
- [x] `match_order` integration: trade + `MatchResult` intact on stats
  overflow, flag observable and round-tripped through snapshot JSON
- [x] Old-format (8-field statistics) package passes `validate()`;
  omit-when-false serde guards
- [x] Pre-push checks clean locally

495 tests passed; 0 failed (476 lib + 9 + 1 + 9 doctests).

## Checklist

- [x] Code follows `rules/global_rules.md` and `CLAUDE.md`
- [x] No `.unwrap()` / `.expect()` / unchecked indexing in production code
- [x] Checked arithmetic — no `saturating_*` / `wrapping_*`
- [x] `tracing` only (WARN per the logging taxonomy)
- [x] Snapshot round-trip + previous-format compatibility test
- [x] No new production `unsafe`, no new dependencies

Closes #117

- **Blocks:** #130

### CODE REVIEW

**Review by joaquinbejar — COMMENTED** [review]
This change fixes the original partial update path, but the reset race, unchecked execution count, unchanged v2 wire identifier, and repeated warning path need correction before merge. The intended effective verdict is REQUEST_CHANGES. GitHub requires COMMENT because this is a self-review.

File: src/price_level/statistics.rs [hunk 1]
```diff
@@ -163,44 +193,81 @@ impl PriceLevelStatistics {
...
+            self.orders_executed.fetch_sub(1, Ordering::Relaxed);
+            self.stats_degraded.store(true, Ordering::Relaxed);
+            return Err(err);
+        }
+
+        if let Err(err) =
+            Self::checked_fetch_add_u64(&self.value_executed, value_u64, "value_executed")
+        {
+            self.quantity_executed
+                .fetch_sub(quantity, Ordering::Relaxed);
```
Comment (joaquinbejar) [inline src/price_level/statistics.rs:419]: 🟠 **Warning**: This rollback can race the public `reset(&self)`, which zeroes the same `Arc`-exposed counter. If reset lands before this `fetch_sub`, the counter wraps toward `u64::MAX`; documenting quiescence does not enforce it for this safe concurrent API. Require exclusive reset access or coordinate reset and recorders with an epoch or active-writer protocol.

File: src/price_level/statistics.rs [hunk 2]
```diff
@@ -163,44 +193,81 @@ impl PriceLevelStatistics {
...
-                message: "value_executed exceeds u64 storage".to_string(),
-            })?;
+        };
 
-        // All fallible validation passed — now apply the mutations. (The only
-        // residual failure mode is a counter nearing `u64::MAX`, inherent to
-        // independent lock-free atomics.)
+        // Commit the additive aggregates with a rollback of the already-committed
+        // prefix on a later overflow (all-or-nothing). `orders_executed` is a
+        // `usize` counter that cannot overflow in practice, but it is part of the
```
Comment (joaquinbejar) [inline src/price_level/statistics.rs]: 🟠 **Warning**: `orders_executed` can overflow because `FromStr` can seed it at `usize::MAX`, after which the unchecked `fetch_add` wraps to zero while the other aggregates advance and this call returns `Ok`. Use a checked `AtomicUsize` CAS and include it in the rollback transaction, with a boundary test.

File: src/price_level/level.rs [hunk 3]
```diff
@@ -727,12 +1002,33 @@ impl PriceLevel {
...
                             data.consumed,
                             data.maker_price,
                             data.maker_timestamp,
                             timestamp.as_u64(),
-                        );
+                        ) {
+                            // WARN, not ERROR: the match is not aborted — this is
+                            // a recoverable observability anomaly flagged by the
+                            // sticky degraded flag (the trade is committed).
+                            tracing::warn!(
```
Comment (joaquinbejar) [inline src/price_level/level.rs:1408]: 🟠 **Warning**: Once an aggregate is saturated, every subsequent committed trade takes this WARN path, creating an unbounded hot-path log flood and tail-latency regression. Emit WARN only on the first false-to-true degraded transition or rate-limit subsequent reports.

File: src/price_level/statistics.rs [hunk 4]
```diff
@@ -487,6 +615,9 @@ impl Serialize for PriceLevelStatistics {
             "sum_waiting_time",
             &self.sum_waiting_time.load(Ordering::Relaxed),
         )?;
+        if degraded {
```
Comment (joaquinbejar) [inline src/price_level/statistics.rs:783]: 🔴 **Critical**: A degraded snapshot now adds `stats_degraded` while the package still identifies itself as format v2. The previous v2 statistics visitor rejects unknown fields, so an older v2 reader cannot deserialize this new v2 payload. Conditional omission only preserves new-reader compatibility with old snapshots. Bump the format to v3, retain and test v2 migration, and update the migration guide and generated README.

File: src/price_level/statistics.rs [hunk 5]
```diff
@@ -147,6 +157,26 @@ impl PriceLevelStatistics {
...
+    /// #113 rollbacks: the subtracted units are exactly the units this call
+    /// added, so the undo is commutative with concurrent `record_execution`
+    /// deltas). That call-backed argument assumes no concurrent
+    /// [`reset`](Self::reset) `store(0)` races the rollback — see the quiescence
+    /// contract on `reset`. So a caller never observes a partial contribution in
+    /// the final state. On any failure — a validation error or a counter overflow —
+    /// the sticky [`stats_degraded`](Self::stats_degraded) flag is set: the
+    /// dropped execution is then observable, even though the caller
+    /// (`PriceLevel::match_order`) cannot fail the already-committed trade.
+    /// Because these are independent lock-free atomics, a *concurrent* reader may
```
Comment (joaquinbejar) [inline src/price_level/statistics.rs:334]: 🟠 **Warning**: This prefix is not necessarily transient because `snapshot()` clones these atomics independently. A concurrent clone can capture and checksum the in-flight prefix as a permanent snapshot, potentially before `stats_degraded` becomes true. Add a version or in-flight protocol so clone and serialization retry until no recording transaction overlaps.

**Review by joaquinbejar — COMMENTED** [review]
One additional concurrency gap remains after the earlier exact-head review because an in-flight statistics prefix can be persisted as a stable snapshot. The intended effective verdict remains REQUEST_CHANGES.

**Review by joaquinbejar — COMMENTED** [review]
The rebased head leaves the existing reset rollback race, unchecked execution count, warning flood, snapshot format compatibility break, and torn statistics snapshot findings unresolved. The intended effective verdict is REQUEST_CHANGES. GitHub requires COMMENT because this is a self review.

**Review by joaquinbejar — COMMENTED** [review]
The current head still has the existing reset rollback race, unchecked execution count, repeated warning flood, unchanged snapshot format compatibility break, and torn statistics snapshot. The intended effective verdict is REQUEST_CHANGES. GitHub requires COMMENT because this is a self review.

**Review by joaquinbejar — COMMENTED** [review]
The current head fixes the original single recorder partial update, but the existing reset race, unchecked execution count, warning flood, snapshot format compatibility break, and torn statistics snapshot findings remain unresolved. The intended effective verdict is REQUEST_CHANGES. GitHub requires COMMENT because this is a self review.

**Review by joaquinbejar — COMMENTED** [review]
The current head fixes the original single recorder partial update, but the existing reset rollback race, unchecked execution count, warning flood, unchanged v2 compatibility break, and persistable torn statistics snapshot remain unresolved. The intended effective verdict is REQUEST_CHANGES. GitHub requires COMMENT because this is a self review.

**Review by joaquinbejar — COMMENTED** [review]
The degraded statistics signal improves observability, but reset can race rollback into a wrapped counter, orders executed still overflows unchecked, snapshots can persist torn statistics, and the changed payload remains mislabeled as format v2. GitHub requires this self review to be submitted as COMMENT, but the intended effective verdict is REQUEST_CHANGES.

**Review by joaquinbejar — COMMENTED** [review]
The degraded statistics signal improves observability, but reset can race rollback into a wrapped counter, orders executed still overflows unchecked, snapshots can persist torn statistics, and the changed payload remains mislabeled as format v2. GitHub requires this self review to be submitted as COMMENT, but the intended effective verdict is REQUEST_CHANGES.

File: src/price_level/statistics.rs [hunk 6]
```diff
@@ -163,44 +193,81 @@ impl PriceLevelStatistics {
...
+        {
+            self.value_executed.fetch_sub(value_u64, Ordering::Relaxed);
+            self.quantity_executed
+                .fetch_sub(quantity, Ordering::Relaxed);
+            self.orders_executed.fetch_sub(1, Ordering::Relaxed);
+            self.stats_degraded.store(true, Ordering::Relaxed);
+            return Err(err);
+        }
+
         self.last_execution_time
```
Comment (joaquinbejar) [inline src/price_level/statistics.rs:443]: 🟠 **Warning**: Concurrent successful recorders can make `last_execution_time` move backwards because an older timestamp can reach this store after a newer call has already stored its value. That violates the accessor contract for the most recent execution even though the additive aggregates include both executions. Use a checked `fetch_max` or CAS update and add a barrier test with distinct timestamps.

**Review by joaquinbejar — COMMENTED** [review]
The rewritten head adds the shared topology fixes and all exact head checks pass, but the existing statistics rollback, overflow, logging, snapshot compatibility, and torn read findings remain unresolved, and concurrent recorders can still move the latest execution timestamp backwards. The intended effective verdict is REQUEST_CHANGES. GitHub requires COMMENT because this is a self review.

**Review by joaquinbejar — COMMENTED** [review]
The issue specific partial update and ignored error paths are fixed and all current checks pass, but the six existing statistics threads remain unresolved on this exact head, including rollback safety, checked overflow, snapshot compatibility and consistency, warning behavior, and timestamp monotonicity. The intended effective verdict is REQUEST_CHANGES. GitHub requires COMMENT because this is a self review.

File: src/price_level/statistics.rs (hunk 1 above)
Comment (joaquinbejar) [reply src/price_level/statistics.rs:419]: Enforced now rather than documented, reset participates as a writer in a new statistics seqlock (RAII guard, closes on every early return), so it can never interleave inside a recording transaction, and the counters keep their checked ops as defense (05cda8e).

File: src/price_level/statistics.rs (hunk 2 above)
Comment (joaquinbejar) [reply src/price_level/statistics.rs]: Fixed, orders_executed is a checked CAS now, first in the all-or-nothing transaction and rolled back like the rest, with a usize::MAX boundary test seeded through the decoder (05cda8e).

File: src/price_level/level.rs (hunk 3 above)
Comment (joaquinbejar) [reply src/price_level/level.rs:1408]: Agreed, the WARN now fires only on the false-to-true transition of the sticky flag (compare_exchange picks the witness), a burst of drops logs once and the flag carries the signal from there (05cda8e).

File: src/price_level/statistics.rs (hunk 4 above)
Comment (joaquinbejar) [reply src/price_level/statistics.rs:783]: Right, an old v2 reader with deny-unknown-fields chokes on the 9-field payload. Format bumped to v3 (new packages write v3, validate accepts v2 legacy and rejects v1), with the skip-when-false statistics serialization kept so checksum recomputation stays version-agnostic and legacy v2 checksums still match. Migration guide and README updated, v2 fixture and v3 degraded/non-degraded round-trips tested (05cda8e).

File: src/price_level/statistics.rs (hunk 5 above)
Comment (joaquinbejar) [reply src/price_level/statistics.rs:334]: Closed by the same seqlock, Clone/Serialize/Display retry their multi-field copy until no recording transaction overlaps, so a checksummed snapshot can no longer capture an in-flight prefix. The reader loop is deliberately unbounded, a bounded torn fallback would defeat the guarantee, liveness argued from the writer-serialization contract and the guard's panic-safe Drop (05cda8e).

File: src/price_level/statistics.rs (hunk 6 above)
Comment (joaquinbejar) [reply src/price_level/statistics.rs:443]: Fixed with fetch_max, an out-of-order recorder can never move the latest-execution timestamp backwards, plus a Barrier test with distinct timestamps asserting the max wins (05cda8e).

**Comment by joaquinbejar** [comment]
Thanks for pushing on the statistics guarantees. All six findings closed in 05cda8e, seqlock-consistent reads with enforced reset serialization, checked orders_executed, transition-only WARN, format v3 with v2 migration, and a monotonic execution timestamp.
